src/main/java/io/projectriff/invoker/JavaFunctionInvokerServer.java

```diff
@@ -66,33 +63,24 @@ public JavaFunctionInvokerServer(Function<Flux<?>, Flux<?>> function, Gson mappe
 
 	private Function<Flux<Message<?>>, Flux<Message<?>>> enhance(
 			Function<Flux<Message<?>>, Flux<Message<?>>> function) {
-		return messages -> function.apply(messages.flatMap(m -> push(m)))
-				.flatMap(m -> pop(m)).subscriberContext(setup());
-	}
-
-	private Function<Context, Context> setup() {
-		return ctx -> ctx.put(CONTEXT, new HashMap<String, Object>());
+		AtomicReference<Map<String, Object>> headers = new AtomicReference<>();
+		return messages -> function.apply(messages.map(m -> push(headers, m)))
+				.map(m -> pop(headers, m));
 	}
 
-	private Mono<Message<?>> pop(Message<?> message) {
-		return Mono.subscriberContext().map(ctx -> {
-			Map<String, Object> map = ctx.get(CONTEXT);
-			if (map.isEmpty()) {
-				return message;
-			}
-			Message<?> result = MessageBuilder.fromMessage(message)
-					.copyHeadersIfAbsent(map).build();
-			map.clear();
-			return result;
-		});
+	private Message<?> pop(AtomicReference<Map<String, Object>> headers,
+			Message<?> message) {
+		Map<String, Object> map = headers.getAndSet(Collections.emptyMap());
+		if (map.isEmpty()) {
```

Comment:
@dsyer Isn't this `if` block unnecessary? If it was a performance optimisation, I'm wondering how often the map will be empty. If it's non-empty most of the time, this test could degrade performance. ;-)

Reply:
If you send in one message and the function sends back 1000, then the headers are empty 999 times, so I think it depends on the function. It's not really a very robust exchange at the minute, just scraping through a smoke test with the gateway not getting angry at the reponses.